src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -931,102 +1021,60 @@ public virtual ODataResource CreateResource(SelectExpandNode selectExpandNode, R
-                        {
-                            AddODataAnnotations(property.InstanceAnnotations, resourceContext, annotation);
-                        }
-                    }                    
-                } 
+                delta = resourceContext.ResourceInstance as IDelta;
             }
-        }
-
-        private void AddODataAnnotations(ICollection<ODataInstanceAnnotation> InstanceAnnotations, ResourceContext resourceContext, KeyValuePair<string, object> annotation)
-        {
-            ODataValue annotationValue = null;
 
-            if (annotation.Value != null)
+            if (delta != null)
             {
-                IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annotation.Value,
-                                            annotation.Value.GetType());
+                if (instanceAnnotationInfo != null)
+                {
+                    delta.TryGetPropertyValue(instanceAnnotationInfo.Name, out instanceAnnotations);
 
-                ODataEdmTypeSerializer edmTypeSerializer = GetEdmTypeSerializer(edmTypeReference);
+                }
+
+                IDeltaSetItem deltaitem = resourceContext.ResourceInstance as IDeltaSetItem;
 
-                if (edmTypeSerializer != null)
+                if (deltaitem != null)
                 {
-                    annotationValue = edmTypeSerializer.CreateODataValue(annotation.Value, edmTypeReference, resourceContext.SerializerContext);
+                    transientAnnotations = deltaitem.TransientInstanceAnnotationContainer;
                 }
             }
             else
             {
-                annotationValue = new ODataNullValue();
-            }
-
-            if (annotationValue != null)
-            {
-                InstanceAnnotations.Add(new ODataInstanceAnnotation(annotation.Key, annotationValue));               
-            }            
-        }
+                if (instanceAnnotationInfo == null || structuredObject == null ||
+                    !structuredObject.TryGetPropertyValue(instanceAnnotationInfo.Name, out instanceAnnotations) || instanceAnnotations == null)
+                {
+                    edmEntityObject = structuredObject as EdmEntityObject;
 
-        private ODataEdmTypeSerializer GetEdmTypeSerializer(IEdmTypeReference edmTypeReference)
-        {
-            ODataEdmTypeSerializer edmTypeSerializer;
-           
-            if (edmTypeReference.IsCollection())
-            {
-                edmTypeSerializer = new ODataCollectionSerializer(SerializerProvider, true);
-            }
-            else if (edmTypeReference.IsStructured())
-            {
-                edmTypeSerializer = new ODataResourceValueSerializer(SerializerProvider);
-            }           
-            else
-            {
-                edmTypeSerializer = SerializerProvider.GetEdmTypeSerializer(edmTypeReference);
+                    if (edmEntityObject != null)
```

Comment:
Lines 1065-1067 can be replaced by:
```
                    if (structuredObject is EdmEntityObject edmEntityObject)
```
No reason to declare `edmEntityObject` in the outer scope...